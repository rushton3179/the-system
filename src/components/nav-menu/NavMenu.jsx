import React from "react";
import PropTypes from "prop-types";
import { faBars } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import "./NavMenu.scss";

class NavMenu extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            dropdownVisible: false
        };

        this.onDropdownClick = this.onDropdownClick.bind(this);
        this.dropDown = React.createRef();
    }

    onDropdownClick() {
        this.setState(prevState => ({
            dropdownVisible: !prevState.dropdownVisible
        }));
    }

    getTransformStyle() {
        const { dropdownVisible } = this.state;
        const dropdownRef = this.dropDown.current;

        if (!dropdownRef) {
            return null;
        }

        if (dropdownVisible) {
            return {
                transform: `translateX(calc(-8vw - ${dropdownRef.offsetWidth}px))`
            };
        }

        return {
            transform: `translateX(0px)`
        };
    }

    renderPointOfInterest(poi) {
        const scrollOptions = {
            behavior: "smooth",
            block: "center",
            inline: "center"
        };
        const onSelect = () => poi.ref.current.scrollIntoView(scrollOptions);

        return (
            <div
                className="the-system-nav-dropdown-goto-item"
                role="button"
                key={poi.display}
                onClick={onSelect}
                onKeyPress={onSelect}
                tabIndex="0"
            >
                {poi.display}
            </div>
        );
    }

    renderPointsOfInterest() {
        const { pointsOfInterest } = this.props;

        return Object.entries(pointsOfInterest).map(poi => this.renderPointOfInterest(poi[1]));
    }

    render() {
        const style = this.getTransformStyle();
        return (
            <div className="the-system-nav">
                <div className="the-system-nav-header">
                    <div
                        className="the-system-nav-header-button"
                        onClick={this.onDropdownClick}
                        onKeyPress={this.onDropdownClick}
                        role="button"
                        tabIndex="0"
                    >
                        <FontAwesomeIcon icon={faBars} />
                    </div>
                </div>
                <div className="the-system-nav-dropdown" ref={this.dropDown} style={style}>
                    <div className="the-system-nav-dropdown-goto-header">Go to</div>
                    {this.renderPointsOfInterest()}
                </div>
            </div>
        );
    }
}

NavMenu.propTypes = {
    pointsOfInterest: PropTypes.shape({}).isRequired
};

export default NavMenu;
