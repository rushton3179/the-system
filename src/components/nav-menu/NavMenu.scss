.the-system-nav {
    animation: fadeIn 2s linear;
    color: white;
    font-family: "Raleway", sans-serif;
    position: fixed;
    right: 2vw;
    text-align: right;
    top: 2vh;
    z-index: 2;

    .the-system-nav-button {
        &:focus {
            outline: none;
        }

        &:hover {
            color: lightblue;
            cursor: pointer;
        }
    }

    .the-system-nav-header-button {
        font-size: 1.8em;
    }

    .the-system-nav-dropdown {
        left: 10vw;
        overflow-y: scroll;
        position: absolute;
        transition: transform 1s;

        .the-system-nav-dropdown-subheader {
            font-size: 1.5em;
            font-weight: bolder;
        }

        .the-system-nav-goto-body {
            overflow-y: hidden;

            &.the-system-nav-goto-body-open {
                max-height: 400px;
                transition: max-height 1s linear;
            }

            &.the-system-nav-goto-body-closed {
                max-height: 0;
                transition: max-height 1s linear;
            }

            .the-system-nav-goto-item {
                font-size: 1.2em;
                padding-left: 10px;
            }
        }

        .the-system-nav-info-body {
            overflow-y: hidden;

            &.the-system-nav-info-body-open {
                max-height: 400px;
                transition: max-height 1s linear;
            }

            &.the-system-nav-info-body-closed {
                max-height: 0;
                transition: max-height 1s linear;
            }

            .the-system-nav-info-body-heading {
                font-weight: bold;
                white-space: nowrap;
            }

            .the-system-nav-info-body-stat {
                padding-bottom: 5px;
                white-space: nowrap;

                &:last-child {
                    padding-bottom: 10px;
                }
            }
        }
    }
}

@keyframes fadeIn {
    from {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
}
