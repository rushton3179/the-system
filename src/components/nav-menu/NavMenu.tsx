import { faBars } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React, { CSSProperties, FC, useState, useRef, RefObject } from 'react';

import { getDistanceToTop } from '../../utils/DomUtil';
import NavMenuSubsection from './NavMenuSubsection';
import styles from './NavMenu.module.scss';
import InfoMenu from './sub-menus/InfoMenu';
import PointOfInterestButton from './sub-menus/PointOfInterestButton';
import { PointsOfInterestMap } from '../../PointsOfInterest';
import { SystemContext } from '../../SystemContext';
import useClickAndEnterKeyDown from '../../hooks/useClickAndEnterKeydown';

interface Props {
  /** Whether the red orbit lines are visible. */
  orbitsVisible: boolean;
  /** The collection of points of interest. */
  pointsOfInterestMap: PointsOfInterestMap;
  /** The point of interest currently being followed. */
  followedPointOfInterest?: { ref: RefObject<HTMLDivElement> };
  /** Change handler for whether the red orbit lines are visible. */
  onOrbitsVisibleChange(orbitsVisible: boolean): void;
  /** Change handler for normalising km per pixel. */
  onChangeSystemSize(systemSizeContext: SystemContext): void;
  /** A handler for following a point of interest around the solar system. */
  onFollowPointOfInterest(pointOfInterest: { ref: RefObject<HTMLDivElement> }): void;
}

type OpenSubsectionState = 'info' | 'nav' | undefined;

/**
 * A sliding nav menu. When the burger icon is clicked the menu will slide into view. Each section will slide
 * open on clicking the title for it, any other open section will be closed at this point.
 *
 * There are no unit tests for this unfortunately as all the logic is based around extracting details from refs.
 * TODO Browser tests
 */
const NavMenu: FC<Props> = ({
  orbitsVisible,
  pointsOfInterestMap,
  followedPointOfInterest,
  onChangeSystemSize,
  onOrbitsVisibleChange,
  onFollowPointOfInterest,
}) => {
  const [menuVisible, setMenuVisible] = useState(false);
  const [openSubsection, setOpenSubsection] = useState<OpenSubsectionState>();
  const menuRef = useRef<HTMLDivElement>(null);

  const [onMenuClick, onMenuEnter] = useClickAndEnterKeyDown(() => setMenuVisible((oldState) => !oldState));
  const onSubsectionClick = (clicked: 'info' | 'nav') =>
    setOpenSubsection((oldState) => (clicked === oldState ? undefined : clicked));

  const transformStyles: CSSProperties = {};

  if (menuRef.current) {
    const maxHeight = `calc(100vh - ${getDistanceToTop(menuRef.current)}px)`;
    const visibleTransform = `translateX(calc(-8vw - ${menuRef.current.offsetWidth}px))`;
    transformStyles.maxHeight = maxHeight;
    transformStyles.transform = menuVisible ? visibleTransform : 'translateX(0px)';
  }

  return (
    <div data-testid="nav-menu" className={styles.nav}>
      <div className={styles.header}>
        <div className={styles.headerButton} onClick={onMenuClick} onKeyPress={onMenuEnter} role="button" tabIndex={0}>
          <FontAwesomeIcon icon={faBars} />
        </div>
      </div>
      <div className={styles.menu} ref={menuRef} style={transformStyles}>
        <NavMenuSubsection
          title={'Info'}
          isVisible={openSubsection === 'info'}
          onHeaderClick={() => onSubsectionClick('info')}
        >
          <InfoMenu
            isVisible={openSubsection === 'info'}
            orbitsVisible={orbitsVisible}
            onChangeSystemSize={onChangeSystemSize}
            onOrbitsVisibleChange={onOrbitsVisibleChange}
          />
        </NavMenuSubsection>
        <NavMenuSubsection
          title={'Navigation'}
          isVisible={openSubsection === 'nav'}
          onHeaderClick={() => onSubsectionClick('nav')}
        >
          <>
            {Object.values(pointsOfInterestMap).map((poi) => (
              <PointOfInterestButton
                isVisible={openSubsection === 'nav'}
                key={poi.display}
                pointOfInterest={poi}
                isBeingFollowed={poi === followedPointOfInterest}
                onPoiClick={() => onFollowPointOfInterest(poi)}
              />
            ))}
          </>
        </NavMenuSubsection>
      </div>
    </div>
  );
};

export default NavMenu;
