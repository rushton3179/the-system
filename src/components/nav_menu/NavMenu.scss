.the-system-nav {
    color: white;
    font-size: 3ex;
    position: fixed;
    right: 2vw;
    top: 2vh;
    z-index: 2;

    :hover {
        color: lightblue;
        cursor: pointer;
    }

    .the-system-nav-header-button {
        &:focus {
            outline: none;
        }
    }

    .the-system-nav-dropdown {
        $dropdown-width: 50px;

        background-color: rgba(85, 85, 85, 0.7);
        height: 50px;
        margin: 3px 6px 0 -6px;
        position: absolute;
        width: 50px;
        transition: transform 1s;

        &.visible {
            transform: translateX(0);
        }

        &.hidden {
            transform: translateX(57px);
        }
    }
}
