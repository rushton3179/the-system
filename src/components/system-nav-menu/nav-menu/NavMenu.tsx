import { faBars } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import React, { CSSProperties, ReactNode, RefObject, ReactNodeArray } from "react";
import { getDistanceToTop } from "../../../utils/DomUtil";
import { getOnEnterPress } from "../../../utils/EventUtils";
import NavMenuSubsection from "./NavMenuSubsection";
import { CallbackFunction } from "../../../types";
import styles from "./NavMenu.module.scss";
import commonStyles from "../CommonStyles.module.scss";

interface Props {
    children: ReactNodeArray;
    titles: ReactNodeArray;
}

interface State {
    menuVisible: boolean;
    openMenuIndex?: number;
}

class NavMenu extends React.PureComponent<Props, State> {
    constructor(props: Props) {
        super(props);
        this.state = {
            menuVisible: false,
        };

        this.menuRef = React.createRef<HTMLDivElement>();
    }

    menuRef: RefObject<HTMLDivElement>;

    onMenuClick: CallbackFunction = () => {
        this.setState(
            (prevState: State): State => ({
                ...prevState,
                menuVisible: !prevState.menuVisible,
            })
        );
    };

    onSubsectionClick = (openMenuIndex?: number): void => {
        this.setState(
            (prevState: State): State => ({
                ...prevState,
                openMenuIndex:
                    openMenuIndex === prevState.openMenuIndex ? undefined : openMenuIndex,
            })
        );
    };

    getTransformStyle(): CSSProperties | undefined {
        const { menuVisible } = this.state;
        const { current } = this.menuRef;

        if (current) {
            const maxHeight = `calc(100vh - ${getDistanceToTop(current)}px)`;
            const visibleTransform = `translateX(calc(-8vw - ${current.offsetWidth}px))`;
            return {
                maxHeight,
                transform: menuVisible ? visibleTransform : "translateX(0px)",
            };
        }
    }

    renderSubsection = (item: ReactNode, index: number): ReactNode => {
        const { titles } = this.props;
        const { openMenuIndex } = this.state;
        return (
            <NavMenuSubsection
                key={index}
                onClick={(): void => this.onSubsectionClick(index)}
                header={titles[index]}
                isVisible={index === openMenuIndex}
            >
                {item}
            </NavMenuSubsection>
        );
    };

    render(): ReactNode {
        const { children, titles } = this.props;
        if (children.length !== titles.length) {
            console.error("NavMenu: children and titles props must equal length.");
            return null;
        }
        return (
            <div className={styles.nav}>
                <div className={styles.header}>
                    <div
                        className={`${commonStyles.button} ${styles.headerButton}`}
                        onClick={this.onMenuClick}
                        onKeyPress={getOnEnterPress(this.onMenuClick)}
                        role="button"
                        tabIndex={0}
                    >
                        <FontAwesomeIcon icon={faBars} />
                    </div>
                </div>
                <div className={styles.menu} ref={this.menuRef} style={this.getTransformStyle()}>
                    {children.map(this.renderSubsection)}
                </div>
            </div>
        );
    }
}

export default NavMenu;
