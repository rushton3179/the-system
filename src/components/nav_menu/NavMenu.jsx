import React from "react";
import { faBars } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import "./NavMenu.scss";

class NavMenu extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            dropdownVisible: false
        };

        this.onDropdownClick = this.onDropdownClick.bind(this);
    }

    onDropdownClick() {
        this.setState(prevState => ({
            dropdownVisible: !prevState.dropdownVisible
        }));
    }

    render() {
        const { dropdownVisible } = this.state;
        const visibility = dropdownVisible ? "visible" : "hidden";

        return (
            <div className="the-system-nav">
                <div className="the-system-nav-header">
                    <div
                        className="the-system-nav-header-button"
                        onClick={this.onDropdownClick}
                        onKeyPress={this.onDropdownClick}
                        role="button"
                        tabIndex="0"
                    >
                        <FontAwesomeIcon icon={faBars} />
                    </div>
                </div>
                <div className={`the-system-nav-dropdown ${visibility}`} />
            </div>
        );
    }
}

export default NavMenu;
