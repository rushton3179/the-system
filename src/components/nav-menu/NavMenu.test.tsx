import React from 'react';
import { render, screen } from '@testing-library/react';

import NavMenu from './NavMenu';

describe('NavMenu', () => {
  test('it renders', () => {
    render(
      <NavMenu
        orbitsVisible={false}
        pointsOfInterestMap={{ somePoint: { display: 'Some point', ref: React.createRef() } }}
        onChangeSystemSize={jest.fn()}
        onFollowPointOfInterest={jest.fn()}
        onOrbitsVisibleChange={jest.fn()}
      />
    );

    expect(screen.getByTestId('nav-menu')).toBeInTheDocument();
  });

  test('it matches snapshot', () => {
    const { container } = render(
      <NavMenu
        orbitsVisible={false}
        pointsOfInterestMap={{ somePoint: { display: 'Some point', ref: React.createRef() } }}
        onChangeSystemSize={jest.fn()}
        onFollowPointOfInterest={jest.fn()}
        onOrbitsVisibleChange={jest.fn()}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
