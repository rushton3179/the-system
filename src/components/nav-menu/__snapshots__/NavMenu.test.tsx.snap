// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavMenu it matches snapshot 1`] = `
<div
  class="nav"
>
  <div
    class="header"
  >
    <div
      class="headerButton"
      data-testid="nav-menu-button"
      role="button"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-bars fa-w-14 "
        data-icon="bars"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
  <div
    class="menu"
    data-testid="nav-menu"
  >
    <div
      class="title"
      role="button"
      tabindex="0"
    >
      info
    </div>
    <div
      class="content"
      style="max-height: 0;"
    >
      <div
        data-testid="info-menu"
      >
        <div
          class="heading"
        >
          infoMenu.time
        </div>
        <div
          class="stat"
          data-testid="time"
        >
          infoMenu.secondsForOneDay
        </div>
        <div
          class="heading"
        >
          infoMenu.distanceBetweenPlanets
        </div>
        <div
          class="stat"
          data-testid="planet-distance"
        >
          infoMenu.kmPerPixel
        </div>
        <div
          class="heading"
        >
          infoMenu.distanceBetweenPlanetsAndMoons
        </div>
        <div
          class="stat"
          data-testid="satellite-distance"
        >
          infoMenu.kmPerPixel
        </div>
        <div
          class="heading"
        >
          infoMenu.planetSize
        </div>
        <div
          class="stat"
          data-testid="planet-size"
        >
          infoMenu.kmPerPixel
        </div>
        <div
          class="buttonWrapper"
        >
          <div
            aria-label="infoMenu.normaliseDistance"
            class="button"
            role="button"
          >
            infoMenu.normaliseDistance
          </div>
        </div>
        <div
          class="buttonWrapper"
        >
          <div
            aria-label="infoMenu.showOrbits"
            class="button"
            role="button"
          >
            infoMenu.showOrbits
          </div>
        </div>
      </div>
    </div>
    <div
      class="title"
      role="button"
      tabindex="0"
    >
      navigation
    </div>
    <div
      class="content"
      style="max-height: 0;"
    >
      <div
        class="navButtons"
      >
        <div>
          <div
            aria-label="pointsOfInterest.sun"
            class="navItem"
            role="button"
          >
            pointsOfInterest.sun
          </div>
        </div>
        <div>
          <div
            aria-label="pointsOfInterest.mercury"
            class="navItem"
            role="button"
          >
            pointsOfInterest.mercury
          </div>
        </div>
        <div>
          <div
            aria-label="pointsOfInterest.venus"
            class="navItem"
            role="button"
          >
            pointsOfInterest.venus
          </div>
        </div>
        <div>
          <div
            aria-label="pointsOfInterest.earth"
            class="navItem"
            role="button"
          >
            pointsOfInterest.earth
          </div>
        </div>
        <div>
          <div
            aria-label="pointsOfInterest.mars"
            class="navItem"
            role="button"
          >
            pointsOfInterest.mars
          </div>
        </div>
        <div>
          <div
            aria-label="pointsOfInterest.theBelt"
            class="navItem"
            role="button"
          >
            pointsOfInterest.theBelt
          </div>
        </div>
        <div>
          <div
            aria-label="pointsOfInterest.jupiter"
            class="navItem"
            role="button"
          >
            pointsOfInterest.jupiter
          </div>
        </div>
        <div>
          <div
            aria-label="pointsOfInterest.saturn"
            class="navItem"
            role="button"
          >
            pointsOfInterest.saturn
          </div>
        </div>
        <div>
          <div
            aria-label="pointsOfInterest.uranus"
            class="navItem"
            role="button"
          >
            pointsOfInterest.uranus
          </div>
        </div>
        <div>
          <div
            aria-label="pointsOfInterest.neptune"
            class="navItem"
            role="button"
          >
            pointsOfInterest.neptune
          </div>
        </div>
        <div>
          <div
            aria-label="pointsOfInterest.pluto"
            class="navItem"
            role="button"
          >
            pointsOfInterest.pluto
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
